### PR TITLE
Generalized `DataConverter` 

### LINKED ISSUE #634: Refactor DataConverter such that it generalizes to other experiments
In our current implementation of `DataConverter`, we assume that the input is i3 files and provide no way for a user to overwrite that assumption. This makes it difficult for other experiments to integrate their file reading and extraction for data conversion. This has been a challenge for @kaareendrup  and recently Jorge from km3net. I suggest that we refactor `DataConverter` under the philosophy: 

>  **_Converting an arbitrary file type to a graphnet-supported backend should be easy._**

To achieve this, I think we can refactor DataConverter into modularized parts:

- *FileReader*: A generic class that is in charge of reading a single file and applying `Extractors` to it. It must return a dictionary where each named field correspond to the output of an extractor. E.g. `data = {'mc_truth' : {'energy': 11.2, 'zenith': 57 .. }`. This dictionary form is the "interim-format" . This means that `Extractors` are assumed to work on only their corresponding `FileReader`.
-  *FileSaveMethod*: A generic class that saves the "interim-format" from `FileReader` to disk. I.e. to parquet or SQLite.
- *DataConverter*: A finalized class (i.e does not need to be implemented by the user) that does all the heavy lifting. It will take as argument `FileReader` and `FileSaveMethod` and tie these two together. It will be in charge of multiprocessing calls to `FileReader` and `event_no` assignment.

A quick diagram illustrates the dataflow below:
image

This refactor would lower the "technical price" of implementing new input file types/sources  data into GraphNeT significantly, as it will only require knowledge on how to open a file and extract the quantities of interest. "What you get for your trouble" is multiprocessing, consistent `event_no` assignment and the ability to save it to any of the implemented `FileSaveMethod`s. I think that's a good bargain.

Below is some pseudo-code on how we could structure the modules:

```python
from typing import List, Union, OrderedDict
from abc import abstractmethod
import glob

class GraphNeTFileReader:

    @abstractmethod
    def __call__(self, file_path: str) -> OrderedDict:
        """ This function should contain the logic for opening and applying
        extractors to a single file. Result should be returned as a Dict."""

    @property
    def accepted_file_extensions(self):
        return self._accepted_file_extensions
    
    @property
    def accepted_extractors(self):
        return self._accepted_extractors
    
    def find_files(self, path: Union[str, List[str]]):
        """ A method for identifying input files in a directory given
            by the user. May be overwritten in custom implementations.

            Will be called by DataConverter if a directory is given as
            input."""
        files = []
        for accepted_file_extension in self.accepted_file_extensions:
            files.extend(glob.glob(path + f'/*{accepted_file_extension}'))
        
        # Check that files are OK.
        self._validate_files(files)
        return files
    
    @final
    def set_extractors(self, extractors: List[GraphNeTExtractors]):
        """ Called by DataConverter to set extractors."""
        self._validate_extractors(extractors)
        self._extractors = extractors
    
    @final
    def _validate_extractors(self):
        # assert given extractors are accepted
    
    @final
    def validate_files(self):
        """ Called by DataConverter to validate input file paths."""
        # assert file extensions are accepted
```

```python
import os

class GraphNeTFileSaveMethod:

    @abstractmethod
    def save_file(self, data: OrderedDict, output_file_path) -> None:
        """ Logic for saving `Data` to a specific file format at 
            `output_file_path`"""
        return
        
    @final
    def __call__(self, data: OrderedDict, file_name: str, out_dir: str) -> None:
        output_file_path = os.join(out_dir, file_name, self.file_extension)
        self._save_file(data = data, output_file_path = output_file_path)
        return

    @property
    def file_extension(self) -> str:
        return self._file_extension
```

   
```python
class DataConverter:
    
    def __init__(file_reader: GraphNeTFileReader,
                 save_method: GraphNeTFileSaveMethod,
                 extractors: List[GraphNeTExtractor],
                 num_workers: int = 1) -> None:
        
        # Member Variable Assignment
        self._file_reader = file_reader
        self._save_method = save_method
        self._num_workers = num_workers

        # Set Extractors. Will throw error if extractors are incompatible
        # with reader.
        self._file_reader.set_extractors(extractors)

    @final
    def __call__(self, input_dir: Union[str, List[str]], output_dir: str) -> None:
        # Get the file reader to produce a list of input files
        # in the directory
        input_files = self._file_reader.find_files(input_dir)
        self._launch_jobs(input_files = input_files, output_dir = output_dir)
    
    @final
    def _launch_jobs(self, input_files: Union[List[str], List[I3FileSets]]) -> None:
        """ Multi Processing Logic.

            Spawns worker pool, 
            distributes the input files evenly across workers.
            declare event_no as globally accessible variable across workers.
            starts jobs.

            Will call process_file in parallel."""
    
    @final
    def _process_file(self, file_path: str):
        """ This function is called in parallel"""
        # Read and apply extractors
        data = self._file_reader(file_path = file_path)

        # Assign event_no's to each event in data
        data = self._assign_event_no(data = data)

        # Create output file name
        output_file_name = self._generate_output_file_name(file_path = file_path)

        # Apply save method
        self._save_method(data = data, 
                          file_name = output_file_name)

```

### PR DESCRIPTION
**Pull request by RasmusOrsoe**
This PR refactors `DataConverter` such that it generalizes to other experiments than icecube, and allows for a very easy implementation of new data backends, i.e it closes #634. 

Please note that the many CodeClimate warnings are associated with code that this PR is not introducing. They are mainly concerned with complexity of icecube-related extractors, which we cannot realistically address. 

**This PR looks huge but it's not.** 

The main changes are:

- Extractors (i3 extractors) have been moved to `graphnet.data.extractors.icecube` . In this new structure, future experiments should put extractors under their own folder; e.g.  `graphnet.data.extractors.km3net`. The content of our existing extractors have not changed.
- `I3Extractor` now inherits from a generic base class for extractors called `Extractor` and can be found in `graphnet.data.extractors.extractor.py`
- Functionality in the old `DataConverter` that opens files and apply extractors to it has been moved to the generic base-class `graphnet.data.readers.GraphNeTFileReader`. An i3-specific file reader can be found in `graphnet.data.readers.I3Reader`. Future integrations of experiments should put their readers here. E.g. `KM3NeTROOTReader.py`
- Functionality in the old `DataConverter` that saves the output of extractors has been moved to the generic base-class `graphnet.data.writers.GraphNeTWriter`. SQLite and parquet specific writers can be found in the same directory.
- A folder for pre-configured `DataConverter`s has been added under `graphnet.data.pre_configured` and here I've added an `I3ToSQLiteConverter` and `I3ToParquetConverter` - these are just pre-defined converters that are identical in terms of usage and functionality to our existing `ParquetDataConverter` and `SQLiteDataConverter`. E.g the reader and writer modules are hardcoded.
- I have added two wrapper classes for `I3ToSQLiteConverter` and `I3ToParquetConverter` under `graphnet.data.{parquet,sqlite}` that are named  `ParquetDataConverter` and `SQLiteDataConverter`. These classes allows existing data conversion scripts to function, but will add a deprecation warning to the logs. Once we release graphnet 2.0, these two directories should be deleted.
- I deleted `I3GenericExtractor` because it does not (and has never really) fit well into our data processing. It is also quite slow, and my impression is that it's largely unused. It would require quite a re-write of the module to make it fit, and I am not too keen on that.

**Usage**
Integrating a data source from new experiment requires implementing a new experiment-specific subclass of `GraphNeTFileReader`, which is fairly straight forward, as it entails writing just a single abstract method. E.g : 

```python
from typing import Union, List, OrderedDict


from graphnet.data.readers import GraphNeTFileReader
from graphnet.data.dataclasses import I3FileSet
from graphnet.data.extractors.my_experiment import MyExtractor

class MyCustomReader(GraphNeTFileReader):
    _accepted_file_extensions = ['.my_special_file_type']
    _accepted_extractors = [MyExtractor]


    def __call__(self, file_path: Union[str, I3FileSet]) -> List[OrderedDict]:
        """Open and apply extractors to a single file.

        The `output` must be a list of dictionaries, where the number of events
        in the file `n_events` satisfies `len(output) = n_events`. I.e each
        element in the list is a dictionary, and each field in the dictionary
        is the output of a single extractor.
        """
```

where `MyExtractor` is a data-source specific implementation of the generic base class `Extractor`:
```python
from typing import Any

from graphnet.data.extractors import Extractor

class MyExtractor(Extractor):
    def __init__(self):
        super().__init__(extractor_name = 'MyExtractor')
    def __call__(self, data: Any) -> dict:
        """Extract information from data."""
        raise NotImplementedError
```
This means that `MyCustomReader` uniquely defines the input to the corresponding `MyExtractor` and how they're called; vectorized, serially, etc. The only limitation imposed by us is that the output format of `MyCustomReader` (see docstring).

Likewise, extending the new `DataConverter` to output files in a new data backend is easy. It requires only the implementation of a single abstract method:

```python
from typing import Dict
from graphnet.data.writers import GraphNeTWriter

class MyNewWriter(GraphNeTWriter):
     _merge_dataframes = True # or False
     _file_extension = '.my_backend'
    def _save_file(
        self,
        data: Union[Dict[str, pd.DataFrame], Dict[str, List[pd.DataFrame]]],
        output_file_path: str,
        n_events: int,
    ) -> None:
        """Save the interim data format from a single input file.

        Args:
            data: the interim data from a single input file.
            output_file_path: output file path.
            n_events: Number of events container in `data`.
        """
        raise NotImplementedError
```
the `_save_file` method receives a dictionary on one of two possible forms, controlled by the class variable `_merge_dataframes`. If `_merge_dataframes= True`, `_save_file` receives a single `pd.DataFrame` object per extractor that needs to be saved to disk. If `_merge_dataframes = False`, `_save_file` receives a `pd.DataFrame` object per event per extractor. By using the pandas framework for this part, exporting the data to new formats is incredible easy. E.g. a writer that saves each extractor output as separate  `.csv` files would be:

```python
from typing import Dict
import os
from graphnet.data.writers import GraphNeTWriter

class MyCSVWriter(GraphNeTWriter):
    _file_extension = '.csv'
    _merge_dataframes = True
    def _save_file(
        self,
        data: Union[Dict[str, pd.DataFrame], Dict[str, List[pd.DataFrame]]],
        output_file_path: str,
        n_events: int,
    ) -> None:
        """Save the interim data format from a single input file.

        Args:
            data: the interim data from a single input file.
            output_file_path: output file path.
            n_events: Number of events container in `data`.
        """
        output_basepath = os.path.basename(output_file_path)
        for extractor in data.keys():
            out_path = output_basepath  + f'_{extractor}' + self.file_extension
            data[extractor].to_csv(out_path)
```

### CODE REVIEW

**Review by Aske-Rosted — COMMENTED** [review]
Looks good to me overall, It's a little hard to grasp all the details of where stuff has been moved, but I imagine it makes sense once you start using it. I have written a few comments and will be happy to approve the request once you've had a look at those. (They don't necessarily require changes to the code)

File: src/graphnet/data/extractors/extractor.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+class Extractor(ABC, Logger):
+    """Base class for extracting information from data files.
+
+    All classes inheriting from `Extractor` should implement the `__call__`
+    method, and should return a pure python dictionary on the form
+
+    output = [{'var1: ..,
+                 ... ,
+              'var_n': ..}]
```
[the remark this answers is not in the document]
Comment (RasmusOrsoe) [reply src/graphnet/data/extractors/extractor.py:16]: It should return a single python dictionary. I updated the docstring to make that clearer.

File: src/graphnet/data/extractors/icecube/i3extractor.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        self._calibration: Optional["icetray.I3Frame.Calibration"] = None
+
+        # Base class constructor
+        super().__init__(extractor_name=extractor_name)
+
+    def set_gcd(self, gcd_file: str, i3_file: str) -> None:
+        """Load the geospatial information contained in the GCD-file."""
+        # If no GCD file is provided, search the I3 file for frames containing
+        # geometry (G) and calibration (C) information.
+        gcd = dataio.I3File(gcd_file or i3_file)
```
[the remark this answers is not in the document]
Comment (RasmusOrsoe) [reply src/graphnet/data/extractors/icecube/i3extractor.py]: I've refactored this function to make it clear what it does and why. Bear in mind this PR did not introduce this code or alter it; this is how it's written currently in our repo.

File: src/graphnet/data/extractors/icecube/i3extractor.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        try:
+            g_frame = gcd.pop_frame(icetray.I3Frame.Geometry)
+        except RuntimeError:
+            self.error(
+                "No GCD file was provided and no G-frame was found. Exiting."
+            )
+            raise
+        else:
```
[the remark this answers is not in the document]
Comment (RasmusOrsoe) [reply src/graphnet/data/extractors/icecube/i3extractor.py]: Nopes. See comment above.

**Comment by RasmusOrsoe** [comment]
@Aske-Rosted thank you for your detailed comments! I ended up refactoring `set_gcd` to make it clearer what the function actually does. I think your questions were a great hint that this function was too compact and opaque - so I have expanded the docstring and added multiple comments. 

> It's a little hard to grasp all the details of where stuff has been moved, but I imagine it makes sense once you start using it.

 Yes - it does look like a bit much, but it really isn't.  You can see [here](https://github.com/graphnet-team/graphnet/blob/3f407b07ee7f595c04b1ab91c1ce59314010e5e2/src/graphnet/data/pre_configured/dataconverters.py) how the new `DataConverter` - `GraphNeTWriter` - `GraphNeTFileReader` combination looks like in practice. 

I'm happy to elaborate on any bits of this code @Aske-Rosted

File: src/graphnet/data/pre_configured/dataconverters.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    def __init__(
+        self,
+        gcd_rescue: str,
+        extractors: List[I3Extractor],
+        outdir: str,
+        index_column: str = "event_no",
+        num_workers: int = 1,
+        i3_filters: Union[I3Filter, List[I3Filter]] = None,  # type: ignore
+    ):
+        """Convert I3 files to Parquet.
```
Comment (Aske-Rosted) [inline src/graphnet/data/pre_configured/dataconverters.py:69]: This one snuck past me...

**Review by Aske-Rosted — APPROVED** [review]
With the correction of the final comment I think this is ready for merging.

**Pull request merged**